src/runtime_basic.jl

```diff
@@ -1406,174 +1379,191 @@ function rt_convert2intvec(a)
-        col_idx += 1
-        if col_idx > ncols
-            col_idx = 1
-            row_idx += 1
-            if row_idx > nrows
+    k = 1
+    for i in 1:nrows
+        for j in 1:ncols
+            if k > length(a)
                 break
             end
+            mat[i, j] = a[k]
+            k += 1
         end
     end
-    return SIntMat(mat)
+    return Sintmat(mat, true)
 end
 
+function rt_cast2intmat(a::Sintmat)
+    return a
+end
+
+function rt_cast2intmat(a::Sintmat, nrows::Int, ncols::Int)
+    return rt__cast2intmat(vec(a.value), nrows, ncols)
+end
+
+function rt_cast2intmat(a::Sintvec)
+    return rt__cast2intmat(a.value, length(a.value), 1)
+end
+
+function rt_cast2intmat(a::Sintvec, nrows::Int, ncols::Int)
+    return rt__cast2intmat(a.value, nrows, ncols)
+end
+
+function rt_cast2intmat(a::Sbigintmat)
+    return rt__cast2intmat(map(Int, vec(a.value)), size(a.value)...)
+end
+
+function rt_cast2intmat(a::Sbigintmat, nrows::Int, ncols::Int)
+    return rt__cast2intmat(map(Int, vec(a.value)), nrows, ncols)
+end
+
+
 #### bigintmat
 
-function rt_convert2bigintmat(a::SBigIntMat)
-    return a
+function rt_convert2bigintmat(a::Sbigintmat)
+    return rt_copy_own(a)
 end
 
-function rt_convert2bigintmat(a::Array{BigInt, 2})
-    return SBigIntMat(deepcopy(a))
+function rt_convert2bigintmat(a::Sintmat)
+    return Sbigintmat(map(BigInt, a.value), false)
 end
 
-function rt_convert2bigintmat(a::Tuple{Vararg{Any}}, nrows::Int, ncols::Int)
-    if nrows <= 0 || ncols <= 0
-        rt_error("nrows and ncols must be positive")
-    end
-    mat = zeros(BigInt, nrows, ncols)
-    row_idx = col_idx = 1
-    for i in a
-        mat[row_idx, col_idx] = rt_convert2bigint(i)
-        col_idx += 1
-        if col_idx > ncols
-            col_idx = 1
-            row_idx += 1
-            if row_idx > nrows
-                break
-            end
-        end
-    end
-    return SBigIntMat(mat)
+function rt_cast2bigintmat(a::Sbigintmat)
+    return a
+end
+
+function rt_cast2bigintmat(a::Sintmat)
+    return Sbigintmat(map(BigInt, a.value), true)
 end
 
 #### list
 
-function rt_convert2list(a::SList)
-    return a
+function rt_convert2list(a::Slist)
+    return rt_copy_own(a)
 end
 
-function rt_convert2list(a::SListData)
-    return SList(deepcopy(a))
+function rt_convert2list(a::Union{Int, BigInt, Sproc, Sintvec, Sintmat, Sbigintmat})
+    return Slist(Any[a], rtInvalidRing, 0, nothing, false)
 end
 
-function rt_convert2list(a::SPoly)
-    @warn_check(a.parent.ring_ptr.cpp_object == rt_basering().ring_ptr.cpp_object, "polynomial encountered from a foreign ring")
-    return SList(SListData(Any[a], a.parent, 1, nothing))
+function rt_convert2list(a::Union{Snumber, Spoly, Svector})
+    @warn_check_rings(a.parent, rt_basering(), "object encountered from a foreign ring")
+    return Slist(Any[a], a.parent, 1, nothing, false)
 end
 
 function rt_convert2list(a::STuple)
+    data::Vector{Any} = map(rt_copy_own, a.list)
```

Comment:
I don't know what `map` will return, but if it doesn't return `Vector{Any}`, it will be converted to that type and induce a new array allocation. I'm not sure if perf is important here, but might be better to write:
```suggestion
    data = Any[rt_copy_own(x) for x in a.list]
```

Reply:
WOW. this improves the permute(8) benchmark from 850ms to 550ms. More than 10x faster than singular now.